### PR TITLE
Display coupon-discounted prices in UI

### LINKED ISSUE #2403: Coupons: display calculated prices in dashboard UI
Part of #2158

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #2403 

#### What's this PR do?
Adds logic to contact the coupons API to get the list of coupons, and calculates adjusted prices for each course run.

#### How should this be manually tested?
Create a coupon locally and verify that it adjusts the price of the 'Pay now' button

### CODE REVIEW

**Comment by aliceriot** [comment]
functionality is good! reading code now.

File: static/js/lib/coupon.js [hunk 1]
```diff
@@ -25,28 +25,26 @@ function makeProgramIdLookup<T: HasProgramId>(obj: Array<T>): Map<number, T> {
...
+        let price = calculateRunPrice(
+          run.id, course.id, program.id, priceLookup.get(program.id), couponLookup.get(program.id)
+        );
+        if (price !== null && price !== undefined) {
+          yield [run.id, price];
+        }
+      }
+    }
+  }
+}
```
Comment (aliceriot) [inline static/js/lib/coupon.js]: generator usage is nice!

File: static/js/lib/coupon.js [hunk 2]
```diff
@@ -25,28 +25,26 @@ function makeProgramIdLookup<T: HasProgramId>(obj: Array<T>): Map<number, T> {
   return new Map(obj.map((value: T) => [value.program_id, value]));
```
Comment (aliceriot) [inline static/js/lib/coupon.js]: can we rename the `obj` argument to be `xs` or `arr` or similar? `obj` isn't so appropriate a name for an array imo.

Comment (aliceriot) [inline static/js/lib/coupon.js]: ( just while we're here! )

File: static/js/containers/DashboardPage_test.js [hunk 3]
```diff
@@ -292,9 +330,11 @@ describe('DashboardPage', () => {
         CLEAR_ENROLLMENTS,
         CLEAR_DASHBOARD,
         CLEAR_COURSE_PRICES,
+        CLEAR_COUPONS,
       ], () => {
         ReactDOM.unmountComponentAtNode(div);
       });
     });
   });
+
```
Comment (aliceriot) [inline static/js/containers/DashboardPage_test.js]: blank line

File: static/js/lib/coupon.js (hunk 1 above)
Comment (noisecapella) [inline static/js/lib/coupon.js]: :100:

File: static/js/lib/coupon.js (hunk 2 above)
Comment (noisecapella) [inline static/js/lib/coupon.js]: Sure

**Comment by noisecapella** [comment]
Comments addressed

**Pull request merged**